### PR TITLE
[AIP-49] Add support for OTel Gauges

### PR DESCRIPTION
**Pull request by ferruzzi**
**Overview**

The next phase of AIP-49, adds support for OpenTelemetry Gauges.

Also cleans up some confusing and inconsistent naming from the previous phase(s)

** Testing **
Passes `breeze testing tests` and `breeze static-checks` locally.  'I have added unit tests and these can also be manually tested by launching Breeze with `breeze start-airflow --integration otel` if anyone wants to check it out.  

Screenshot below shows Gauges in action in Grafana.

image

cc: @o-nikolas @vincbeck @vandonr-amz @syedahsn
ping @howardyoo and @potiuk as having shown an interest in the AIP

### CODE REVIEW

File: airflow/metrics/otel_logger.py [hunk 1]
```diff
@@ -55,7 +60,13 @@
 # https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#updowncounter
 UP_DOWN_COUNTERS = {"airflow.dag_processing.processes"}
 
-METRIC_NAME_PREFIX = "airflow."
+DEFAULT_METRIC_NAME_PREFIX = "airflow"
+# Delimiter is placed between the universal metric prefix and the unique metric name.
+DEFAULT_METRIC_NAME_DELIMITER = "."
+
+
+def full_name(name: str, *, prefix: str = DEFAULT_METRIC_NAME_PREFIX) -> str:
```
Comment (ferruzzi) [inline airflow/metrics/otel_logger.py:68]: This used to live in the test module.  I moved it here and use it in both places now.

File: airflow/metrics/otel_logger.py [hunk 2]
```diff
@@ -55,7 +60,13 @@
 # https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#updowncounter
 UP_DOWN_COUNTERS = {"airflow.dag_processing.processes"}
 
-METRIC_NAME_PREFIX = "airflow."
+DEFAULT_METRIC_NAME_PREFIX = "airflow"
+# Delimiter is placed between the universal metric prefix and the unique metric name.
+DEFAULT_METRIC_NAME_DELIMITER = "."
```
Comment (ferruzzi) [inline airflow/metrics/otel_logger.py:65]: We can maybe eventually add support for custom delimiters to the config, but that sounds like an idea for another day.  This should make it relatively easy to add that later.

File: airflow/metrics/otel_logger.py [hunk 3]
```diff
@@ -83,10 +94,45 @@ def name_is_otel_safe(prefix: str, name: str) -> bool:
...
+    :param name: The original metric name
+    :returns: The name, truncated to an OTel-acceptable length if required
+    """
+    otel_safe_name = name[:OTEL_NAME_MAX_LENGTH]
+    if name != otel_safe_name:
+        warnings.warn(
+            f"Metric name `{name}` exceeds OpenTelemetry's name length limit of "
+            f"{OTEL_NAME_MAX_LENGTH} characters and will be truncated to `{otel_safe_name}`."
+        )
+    return otel_safe_name
```
Comment (ferruzzi) [inline airflow/metrics/otel_logger.py:125]: Moved out of `_create_counter` into a helper so `_create_gauge` can reuse it.

**Review by ferruzzi — COMMENTED** [review]
Just left a few comments

File: airflow/metrics/otel_logger.py (hunk 3 above)
Comment (vincbeck) [inline airflow/metrics/otel_logger.py:125]: Super tiny nit: it MIGHT be easier to understand that way

```suggestion
    if len(name) > OTEL_NAME_MAX_LENGTH:
        warnings.warn(
            f"Metric name `{name}` exceeds OpenTelemetry's name length limit of "
            f"{OTEL_NAME_MAX_LENGTH} characters and will be truncated to `{otel_safe_name}`."
        )
    return name[:OTEL_NAME_MAX_LENGTH]
```

File: airflow/metrics/otel_logger.py [hunk 4]
```diff
@@ -154,9 +200,20 @@ def gauge(
         delta: bool = False,
         *,
         tags: Attributes = None,
+        back_compat_name: str = "",
     ) -> None:
-        warnings.warn("OpenTelemetry Gauges are not yet implemented.")
-        return None
+        if rate < 0:
+            raise ValueError("rate must be a positive value.")
+        if rate < 1 and random.random() > rate:
```
Comment (vincbeck) [inline airflow/metrics/otel_logger.py]: Maybe add a comment to explain why?

Comment (ferruzzi) [reply airflow/metrics/otel_logger.py]: Just noticed I missed the docstring for this method.   I'll add that in;  it'll be the same as the incr() and decr() ones.   Do you think it needs more explanation than I have in those?  I can update them all while I'm at it (and fix the typo in those ones)

Comment (vincbeck) [reply airflow/metrics/otel_logger.py]: Oh I see! That should be enough to me

File: airflow/metrics/otel_logger.py (hunk 3 above)
Comment (ferruzzi) [reply airflow/metrics/otel_logger.py:125]: I suppose I could, but I would still need to assign the variable above so I can include it in the warning so really all it would do is change the condition from checking if they are the same to checking if the length fits.  I don't have a strong preference either way, I'm not sure it matters either way.

File: airflow/metrics/otel_logger.py [hunk 5]
```diff
@@ -83,10 +95,45 @@ def name_is_otel_safe(prefix: str, name: str) -> bool:
...
+    """
+    Given an OpenTelemetry Instrument, returns the type of instrument as a string.
+
+    type() will return something like: <class 'opentelemetry.sdk.metrics._internal.instrument._Counter'>
+    This extracts that down to just "Counter" (or "Gauge" or whatever) for cleaner logging purposes.
+
+    :param obj: An OTel Instrument or subclass
+    :returns: The type() of the Instrument without all the nested class info
+    """
+    return str(type(obj)).split(".")[-1][1:-2]
```
Comment (vandonr-amz) [inline airflow/metrics/otel_logger.py]: that's a lot of twisting, wdyt of using `type(obj).__name__[1:]` ?

File: airflow/metrics/otel_logger.py [hunk 6]
```diff
@@ -83,10 +95,45 @@ def name_is_otel_safe(prefix: str, name: str) -> bool:
     return bool(stat_name_otel_handler(prefix, name, max_length=OTEL_NAME_MAX_LENGTH))
 
 
+def _type_as_str(obj: Instrument) -> str:
+    """
+    Given an OpenTelemetry Instrument, returns the type of instrument as a string.
+
+    type() will return something like: <class 'opentelemetry.sdk.metrics._internal.instrument._Counter'>
+    This extracts that down to just "Counter" (or "Gauge" or whatever) for cleaner logging purposes.
```
Comment (vandonr-amz) [inline airflow/metrics/otel_logger.py]: This kind of "technical explanation" is very nice to have, but I think it should be in a `#` comment and not in the docstring

File: airflow/metrics/otel_logger.py [hunk 7]
```diff
@@ -235,6 +301,57 @@ def del_counter(self, name: str, attributes: Attributes = None) -> None:
...
+        :param name: The name of the gauge to record.
+        :param value: The new reading to record.
+        :param delta: If True, value is added to the previous reading, else it overrides.
+        :param tags: Gauge attributes which were used to generate a unique key to store the counter.
+        :returns: None
+        """
+        key: str = _generate_key_name(name, tags)
+        old_value = self.poke_gauge(name, tags)
+        # If delta is true, add the new value to the last reading otherwise overwrite it.
+        self.map[key] = Observation(value + (delta * old_value), tags)
```
Comment (vandonr-amz) [inline airflow/metrics/otel_logger.py]: oh wow because `True == 1` ? And the type checkers don't say anything ? 😱

Comment (ferruzzi) [reply airflow/metrics/otel_logger.py]: Yup, python will happily treat True as 1; None and False are 0.  :)   

Another handy use for that is if you want to see how many Trues you have, you can use sum, like "How many names contain the letter A?" could be `sum(['A' in name for name in names])`  You'll end up with a list[bool] containing both True and False values, and sum() will add all the 1's and 0's accordingly.

File: airflow/metrics/otel_logger.py (hunk 5 above)
Comment (ferruzzi) [reply airflow/metrics/otel_logger.py]: I'll double check that it's the same result, but if it works I'm all for it. :+1:

Comment (ferruzzi) [reply airflow/metrics/otel_logger.py]: Yeah, nice.  This is perfect, thanks!   Also, I'm embarrassed to notice that I never actually used this method.  Updated the method and the spot I had intended to use it.

File: airflow/metrics/otel_logger.py (hunk 6 above)
Comment (ferruzzi) [reply airflow/metrics/otel_logger.py]: That is totally fair, I'll move it.

File: airflow/metrics/otel_logger.py [hunk 8]
```diff
@@ -83,10 +95,45 @@ def name_is_otel_safe(prefix: str, name: str) -> bool:
     return bool(stat_name_otel_handler(prefix, name, max_length=OTEL_NAME_MAX_LENGTH))
 
 
+def _type_as_str(obj: Instrument) -> str:
+    """
+    Given an OpenTelemetry Instrument, returns the type of instrument as a string.
```
Comment (o-nikolas) [inline airflow/metrics/otel_logger.py]: super nit:
```suggestion
    Given an OpenTelemetry Instrument, returns the type of the instrument as a string.
```

File: airflow/metrics/otel_logger.py (hunk 6 above)
Comment (o-nikolas) [inline airflow/metrics/otel_logger.py]:
```suggestion
    This extracts that down to just "Counter" (or "Gauge", etc) for cleaner logging purposes.
```

File: airflow/metrics/otel_logger.py (hunk 7 above)
Comment (o-nikolas) [reply airflow/metrics/otel_logger.py]: While neat, I find this a little obtuse/esoteric. I'd just go with a more straightforward if-statement or ternary operator so that it is more readable for future folks. Or at the very least a comment above this line explaining it.

File: airflow/metrics/otel_logger.py [hunk 9]
```diff
@@ -235,6 +301,57 @@ def del_counter(self, name: str, attributes: Attributes = None) -> None:
         if key in self.map.keys():
             del self.map[key]
 
+    def set_value(self, name: str, value: float, delta: bool, tags: Attributes):
```
Comment (o-nikolas) [inline airflow/metrics/otel_logger.py]: minor, all the other methods are `<something>_gauge`, I feel like this one should follow suit?
```suggestion
    def set_gauge(self, name: str, value: float, delta: bool, tags: Attributes):
```

File: tests/core/test_otel_logger.py [hunk 10]
```diff
@@ -173,13 +171,54 @@ def test_decr_with_rate_limit_works(self, mock_random, name):
...
+
+        self.meter.get_meter().create_observable_gauge.assert_called_once_with(
+            name=full_name(name), callbacks=ANY
+        )
+        assert self.map[full_name(name)].value == 2
+
+    @mock.patch("random.random", side_effect=[0.1, 0.9])
+    @mock.patch.object(MetricsMap, "set_value")
+    def test_gauge_with_rate_limit_works(self, mock_set_value, mock_random, name):
+        # Create the counter and set the value to 1
```
Comment (o-nikolas) [inline tests/core/test_otel_logger.py]:
```suggestion
        # Create the gauge and set the value to 1
```

File: tests/core/test_otel_logger.py [hunk 11]
```diff
@@ -173,13 +171,54 @@ def test_decr_with_rate_limit_works(self, mock_random, name):
         self.map[full_name(name)].add.assert_has_calls(expected_calls)
         self.map[full_name(name)].add.call_count == 2
 
-    @mock.patch("warnings.warn")
-    def test_timer_warns_not_implemented(self, mock_warn):
-        with self.stats.timer():
-            mock_warn.assert_called_once_with("OpenTelemetry Timers are not yet implemented.")
+    def test_gauge_new_metric(self, name):
```
Comment (o-nikolas) [inline tests/core/test_otel_logger.py:174]: Should there be a test for getting the value of a gauge? All the tests below seem to be about setting/creating them.

File: airflow/metrics/otel_logger.py [hunk 12]
```diff
@@ -154,9 +201,33 @@ def gauge(
...
+        :param delta: If true, the provided value will be added to the previous value.
+            If False the new value will override the previous.
+        :param tags: Tags to append to the stat.
+        :param back_compat_name:  If an alternative name is provided, the
+            stat will be emitted using both names if possible.
+        """
+        if rate < 0:
+            raise ValueError("rate must be a positive value.")
+        if rate < 1 and random.random() > rate:
+            return
```
Comment (o-nikolas) [inline airflow/metrics/otel_logger.py]: Probably worth factoring this out into a helper instead of reimplementing it twice (soon to be three with timers)

File: airflow/metrics/otel_logger.py (hunk 9 above)
Comment (ferruzzi) [reply airflow/metrics/otel_logger.py]: Yeah, I guess that's fair.  I'll make the changes.

File: tests/core/test_otel_logger.py (hunk 11 above)
Comment (ferruzzi) [reply tests/core/test_otel_logger.py:174]: I checked it [here](https://github.com/apache/airflow/pull/31725/files#diff-1a02f43043ead15894e9b58e32b1ed9a4a079ed461ab73b66f28d984baeab261R199) in the "test_gauge_existing_metric" test, but I can certainly add that check in to the "test_gauge_new_metric" test as well.  Never hurts to have more confirmation :+1:

Comment (ferruzzi) [reply tests/core/test_otel_logger.py:174]: And I'll add a new test to verify that `delta` is working as intended :+1:    I think that should cover our bases on that?

File: airflow/metrics/otel_logger.py [hunk 13]
```diff
@@ -83,10 +95,51 @@ def name_is_otel_safe(prefix: str, name: str) -> bool:
...
 
+def _type_as_str(obj: Instrument) -> str:
+    """
+    Given an OpenTelemetry Instrument, returns the type of the instrument as a string.
+
+    :param obj: An OTel Instrument or subclass
+    :returns: The type() of the Instrument without all the nested class info
+    """
+    # type() will return something like: <class 'opentelemetry.sdk.metrics._internal.instrument._Counter'>
+    # This extracts that down to just "Counter" (or "Gauge", etc.) for cleaner logging purposes.
```
Comment (vandonr-amz) [inline airflow/metrics/otel_logger.py]: this is out of date with the new code :p
the only thing of interest to mention now is that class names are prefixed by `_`, which is why the `[1:]` is necessary

**Review by vandonr-amz — APPROVED** [review]
looking good to me, with just a little comment to fix

File: airflow/metrics/otel_logger.py (hunk 7 above)
Comment (ferruzzi) [reply airflow/metrics/otel_logger.py]: Alright, I've changed it to an if block to make it more obvious.

**Review by o-nikolas — APPROVED** [review]

File: airflow/metrics/otel_logger.py (hunk 13 above)
Comment (ferruzzi) [reply airflow/metrics/otel_logger.py]: fixed, thanks.    I'd say it's still kind of accurate, just less relevant :P

File: airflow/metrics/otel_logger.py [hunk 14]
```diff
@@ -100,22 +153,22 @@ def incr(
         rate: float = 1,
         tags: Attributes = None,
     ):
-        """
+        """counter
```
Comment (jedcunningham) [inline airflow/metrics/otel_logger.py]: Was this change intentional?

File: airflow/metrics/otel_logger.py [hunk 15]
```diff
@@ -166,7 +241,7 @@ def timing(
         *,
         tags: Attributes = None,
     ) -> None:
-        warnings.warn("OpenTelemetry Timers are not yet implemented.")
+        warnings.warn(f"Create timer {stat} : OpenTelemetry Timers are not yet implemented.")
```
Comment (jedcunningham) [inline airflow/metrics/otel_logger.py]:
```suggestion
        warnings.warn(f"Create timer {stat}: OpenTelemetry Timers are not yet implemented.")
```

tiny nit

File: airflow/metrics/otel_logger.py [hunk 16]
```diff
@@ -177,7 +252,7 @@ def timer(
         tags: Attributes = None,
         **kwargs,
     ) -> TimerProtocol:
-        warnings.warn("OpenTelemetry Timers are not yet implemented.")
+        warnings.warn(f"Create timer {stat} : OpenTelemetry Timers are not yet implemented.")
```
Comment (jedcunningham) [inline airflow/metrics/otel_logger.py]:
```suggestion
        warnings.warn(f"Create timer {stat}: OpenTelemetry Timers are not yet implemented.")
```

Another tiny nit

**Review by vincbeck — APPROVED** [review]

File: airflow/metrics/otel_logger.py [hunk 17]
```diff
@@ -235,6 +304,60 @@ def del_counter(self, name: str, attributes: Attributes = None) -> None:
...
+    def poke_gauge(self, name: str, attributes: Attributes = None) -> GaugeValues:
+        """
+        Returns the value of the gauge; creates a new one with the default value if it did not exist.
+
+        :param name: The name of the gauge to fetch or create.
+        :param attributes:  Gauge attributes, used to generate a unique key to store the gauge.
+        :returns:  The integer or float value last recorded for the provided Gauge name.
+        """
+        key = _generate_key_name(name, attributes)
+        if key not in self.map.keys():
```
Comment (jedcunningham) [inline airflow/metrics/otel_logger.py]:
```suggestion
        if key not in self.map:
```

nit

**Review by jedcunningham — APPROVED** [review]

File: airflow/metrics/otel_logger.py (hunk 14 above)
Comment (ferruzzi) [reply airflow/metrics/otel_logger.py]: No, thanks for the catch.  I'll fix it.

**Comment by ferruzzi** [comment]
Thanks @jedcunningham , [this](https://github.com/apache/airflow/pull/31725/commits/4f843bc38b2b9b16a405979504b46eaeb744c0c4) should incorporate your suggestions.

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Looks cool
